src/librbd/mirror/snapshot/CreatePrimaryRequest.cc

```diff
@@ -115,6 +115,9 @@ void CreatePrimaryRequest<I>::create_snapshot() {
       cls::rbd::MIRROR_SNAPSHOT_STATE_PRIMARY_DEMOTED :
       cls::rbd::MIRROR_SNAPSHOT_STATE_PRIMARY),
     m_mirror_peer_uuids, "", m_clean_since_snap_id};
+  if (m_snap_id == nullptr) {
+    ns.complete = true;
```

Comment:
This is wrong.  If you look at the callers of `CreatePrimaryRequest`, `m_snap_id == nullptr` is just an indicator of whether the caller is interested in the ID of the snapshot that is created or not and this is exposed all the way up to the public API (see `rbd_mirror_image_create_snapshot` signature, for example).

What this would effectively do is prematurely mark _any_ mirror snapshot for which a user happens to not care about the ID as complete.

Reply:
> Currently snapshots created while promote operation are not marked complete.

Taking a step back, do you an output or a log snippet to back this up?  If have just tried a "relocate" operation (`rbd mirror image demote` on one cluster followed by `rbd mirror image promote` on another cluster) and both demotion and promotion snapshots appear to be complete:
```
    {
        "id": 7,
        "name": ".mirror.primary.4cacd615-32ab-46fc-a041-fe1b4a05705e.c318dfbc-ca9f-4ff9-bbb8-db94d1a9bcb7",
        "size": 1073741824,
        "protected": "false",
        "timestamp": "Thu Apr 13 15:21:59 2023",
        "namespace": {
            "type": "mirror",
            "state": "demoted",
            "mirror_peer_uuids": [
                "43872b76-65bb-4330-a59c-b0302245f12d"
            ],
            "complete": true
        }
    },
```
```
    {
        "id": 9,
        "name": ".mirror.primary.4cacd615-32ab-46fc-a041-fe1b4a05705e.e2a328ba-4103-40c5-b3cb-4e6d0e94deef",
        "size": 1073741824,
        "protected": "false",
        "timestamp": "Thu Apr 13 15:22:21 2023",
        "namespace": {
            "type": "mirror",
            "state": "primary",
            "mirror_peer_uuids": [
                "746b1cc4-b27e-4fa4-b4bc-7c62ff74c7b1"
            ],
            "complete": true
        }
    }
```

Reply:
As part of the promote we create a snapshot so we land in `CreatePrimaryRequest<I>::create_snapshot()  {} `, once snap create is done we will eventually we end up in `CreatePrimaryRequest<I>::handle_create_snapshot(int r){}`

`handle_create_snapshot()` will call refresh_image();.
Note `m_snap_id = null` at this point and `complete=false`

```
template <typename I>                                                   
void CreatePrimaryRequest<I>::refresh_image() {                         
  // if snapshot created via remote RPC, refresh is required to retrieve
  // the snapshot id                                                    
  CephContext *cct = m_image_ctx->cct;                                  
  ldout(cct, 15) << dendl;                                              
  if (m_snap_id == nullptr) {                                           
  ┆  ┆ unlink_peer();                                                      
  ┆ return;                                                             
  }
  [...]
}                                                                     
```

Since `complete=false` in the unlink_peer() we will attempt to delete the snapshot.

Reply:
@idryomov 

@I

> What this would effectively do is prematurely mark any mirror snapshot for which a user happens to not care about the ID as complete.

In addition to `if (m_snap_id == nullptr){}` use in `CreatePrimaryRequest<I>::refresh_image()`, I also see `CreateNonPrimaryRequest<I>::create_snapshot()` doing ` ns.complete = is_orphan();`

Where:
```
bool is_orphan() const {               
┆ return m_primary_mirror_uuid.empty();
}                                      
```
m_primary_mirror_uuid is the 3rd argument to `CreateNonPrimaryRequest<I>::create()`

```
[0] pkalever 😎 ceph✨ git grep -n "CreateNonPrimaryRequest" | grep "::create("
src/librbd/mirror/snapshot/PromoteRequest.cc:62:  auto req = CreateNonPrimaryRequest<I>::create(
src/tools/rbd_mirror/image_replayer/snapshot/Replayer.cc:972:  auto req = librbd::mirror::snapshot::CreateNonPrimaryRequest<I>::create(
```

Checking only PromoteRequest.cc here, as that's what we care:

```
void PromoteRequest<I>::create_orphan_snapshot() {
[...]
  auto req = CreateNonPrimaryRequest<I>::create(
  ┆ m_image_ctx, false, "", CEPH_NOSNAP, {}, {}, nullptr, ctx);
  req->send();
}
```

Should we also consider the CreatePrimaryRequest:create_snapshot() comes as part of the promote operation as a orphan snapshot and do the same magic?

I don't see any other better option!


Thanks!

Reply:
> Since `complete=false` in the unlink_peer() we will attempt to delete the snapshot.

Then the original fix is also wrong and the logic in `CreatePrimaryRequest::unlink_peer` needs to be adjusted a bit more.  The desired behavior is as follows: after creating `snap2` (current snapshot), `snap1` (previous snapshot that is expected to have been marked complete by now) should be removed if it's still incomplete.

Your description makes it sound like there is a case where `snap2` (current snapshot, i.e. one that has just been created and is obviously incomplete because the associated image state data hasn't been written out yet) is removed.  That should never happen -- the current snapshot should be allowed though to give it a chance to get marked complete later, outside of `CreatePrimaryRequest`.

Reply:
> I also see `CreateNonPrimaryRequest<I>::create_snapshot()` doing ` ns.complete = is_orphan();`

This is not relevant as we are only concerned with primary snapshots here.

Reply:
I had updated it to skip UnlinkPeerRequest for recent mirror snapshot. PTAL